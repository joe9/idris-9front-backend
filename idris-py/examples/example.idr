module Main

import Python
import Python.Prim
import Python.Exceptions

-- These modules contain signatures for Python libraries.
import Python.Lib.Os
import Python.Lib.Requests
import Python.Lib.BeautifulSoup
import Python.Lib.Queue
import Python.Lib.Threading

%default total

-- Even though field names are strings,
-- everything is typechecked according to the signatures imported above.

partial
main : PIO ()
main = do
  reqs <- Requests.import_

  -- (/) extracts the named attribute
  -- ($) calls a function
  -- (/.) and ($.) work with pure LHS
  -- (/:) and ($:) work with monadic LHS (useful for chaining)
  --
  -- equivalent to: session = reqs.Session()
  session <- reqs /. "Session" $. []

  -- equivalent to: html = session.get("http://idris-lang.org").text
  -- Notice that chaining is not a problem.
  html <- session /. "get" $. ["http://idris-lang.org"] /: "text"

  -- import Beautiful Soup
  bs4 <- BeautifulSoup.import_

  -- construct soup from HTML
  soup <- bs4 /. "BeautifulSoup" $. [html, Parsers.HTML]

  -- get the iterator over <li> elements, given by CSS selector
  features <- soup /. "select" $. ["div.entry-content li"]

  -- print all <li> elements as features
  putStrLn' $ "Idris has got the following exciting features:"
  count <- iterate features 0 $ \i : Int, li : Obj Element => do
    -- collect : Iterator a -> PIO (List a)
    line <- concat <$> collect (li /. "strings")
    putStrLn' $ show (i+1) ++ ". " ++ line
    pure $ i + 1

  putStrLn' $ "Total number of features: " ++ show count
  putStrLn' ""


  -- ###  Concurrency  ###

  let thread : (String -> PIO Nat) = \name => do
    putStrLn' $ "thread " ++ name ++ " starting"
    html <- session /. "get" $. ["http://idris-lang.org"] /: "text"
    putStrLn' $ "thread " ++ name ++ " done"
    pure $ length html

  thrA <- forkPIO $ thread "A"
  thrB <- forkPIO $ thread "B"
  resA <- wait thrA
  resB <- wait thrB

  putStrLn' $ "thread A says " ++ show resA
  putStrLn' $ "thread B says " ++ show resB
  putStrLn' ""


  -- ###  Exceptions  ###

  os <- Os.import_
  putStrLn' "And now, let's fail!"

  -- the standard try-catch variant
  try (do
    os /. "mkdir" $. ["/root/hello"]
    putStrLn' $ "Something's wrong, your root's homedir is writable!"
  ) `catch` (\etype, e => case etype of
    OSError => putStrLn' $ "  -> (1) everything's fine: " ++ showException e
    _       => raise e
  )

  -- Idris sugar, originally used in Effects
  OK ret <- try $ os /. "mkdir" $. ["/root/hello"]
    | Except OSError e => putStrLn' ("  -> (2) everything's fine: " ++ showException e)
    | Except _       e => raise e
  putStrLn' $ "Your root could probably use some security lessons!"

exports : FFI_Export FFI_Py "example.py" []
exports =
    Fun greet "greet" $
    End
  where
    greet : String -> PIO ()
    greet name = putStrLn' $ "Hello " ++ name ++ "!"
