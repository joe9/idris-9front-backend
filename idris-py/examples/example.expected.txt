Idris has got the following exciting features:
1. Full dependent types with dependent pattern matching
2. Simple foreign function interface (to C)
3. Compiler-supported interactive editing: the compiler helps you write code using the types
4. where clauses, with rule, simple case expressions, pattern matching let and lambda bindings
5. Dependent records with projection and update
6. Interfaces (similar to type classes in Haskell)
7. Type-driven overloading resolution
8. do notation and idiom brackets
9. Indentation significant syntax
10. Extensible syntax
11. Cumulative universes
12. Totality checking
13. Hugs style interactive environment
Total number of features: 13

thread A starting
thread B starting
thread A or B done
thread A or B done
thread A says 11024
thread B says 11024

And now, let's fail!
  -> (1) everything's fine: [Errno 13] Permission denied: '/root/hello'
  -> (2) everything's fine: [Errno 13] Permission denied: '/root/hello'
